middleware/hosts/hosts.go

```diff
@@ -0,0 +1,178 @@
+package hosts
+
+import (
+	"errors"
+	"net"
+
+	"golang.org/x/net/context"
+
+	"github.com/coredns/coredns/middleware"
+	"github.com/coredns/coredns/middleware/pkg/dnsutil"
+	"github.com/coredns/coredns/request"
+	"github.com/miekg/dns"
+)
+
+// Hosts is the middleware handler
+type Hosts struct {
+	Next middleware.Handler
+	*Hostsfile
+
+	Origins     []string
+	Fallthrough bool
+}
+
+// ServeDNS implements the middleware.Handle interface.
+func (h Hosts) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	state := request.Request{W: w, Req: r}
+	if state.QClass() != dns.ClassINET {
+		return dns.RcodeServerFailure, middleware.Error(h.Name(), errors.New("can only deal with ClassINET"))
+	}
+	qname := state.Name()
+
+	answers := []dns.RR{}
+
+	// Precheck with the origins, i.e. are we allowed to looks here.
+	if h.Origins != nil {
```

Comment:
My assumption was if you are specifying the zones in the corefile, and the query is for a zone that is not in that list, then we want to failover to the next middleware even if failover is not specified. Is that assumption correct, or is there another way to handle that case?

Or, as I just realized may happen, does the c.ServerBlockKeys call return a zone regardless, "." perhaps? I'll test that out later tonight.